routing/opening_hours_serdes.hpp

```diff
@@ -0,0 +1,486 @@
+#pragma once
+
+#include "coding/bit_streams.hpp"
+#include "coding/varint.hpp"
+
+#include "base/checked_cast.hpp"
+#include "base/stl_helpers.hpp"
+
+#include <limits>
+#include <set>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "3party/opening_hours/opening_hours.hpp"
+
+// TODO (@gmoryes)
+//  opening_hours does not support such string: 2019 Apr - 2020 May,
+//  but 2019 Arp 14 - 2020 May 31 supports.
+
+namespace routing
+{
+/// \brief This class serialise opening hours from OSM format to some binary data.
+/// Each opening hours has one, two, ... rules. Each rule has some data, like: Year, Month,
+/// Weekdays etc. (look to |Header::Bits| for supported features). This class has two mods:
+///     1) (default) Serialize opening hours only if all rules have same headers. In this case
+///        we serialize in such format:
+///        <number of rules><header><rule 1>...<rule N>
+///     2) (Turn on with |SetSerializeEverything| method) For each rule serialize its' header.
+///        Format:
+///        <number of rules><header 1><rule 1><header 2><rule 2>...<header N><rule N>
+///
+/// First format can be used in routing for access:conditional because it has more strong filtering
+/// of opening hours and the more trash will be dropped.
+/// Second format can be used in opening hours for POI because it support more kinds of opening
+/// hours and in case of invalid data we can show raw string and user will able to understand what
+/// it's supposed to be.
```

Comment:
а где тут про то что мы игнорим правила в которые входят годы доо 2020 и правила, которые заканчиваются в году, который уже прошёл на момент сериализации?

Reply:
да, то из-за чего писал и забыл)
```
/// This class also drops rules with old year range, like: 
///     "2010 - 2011"
///     "2010 Arp 19 - 2014 May 20"
///     etc.
/// In case of mixed data: 
///     "2010 Arp 19 - 2014 May 20 ; 2020 Mo-Su"
///     "2010 Arp 19 - 2014 May 20 ; Mo-Su 10:00-15:00"
/// It leaves only fresh rules:
///     "2020 Mo-Su" and "Mo-Su 10:00-15:00"
```